File: MapEditingTools/MapEditingTools/Quaternion.js
function toRadians(angle)
{
  return (angle % 360) * (Math.PI/180);
}

class quaternion
{
  constructor()
  {
    switch(arguments.length)
    {
      case 0:
        this.W = 1;
        this.X = 0;
        this.Y = 0;
        this.Z = 0;
        
        break;
      
      case 3:
        var cosx = Math.cos(toRadians(arguments[0]) * 0.5);
        var cosy = Math.cos(toRadians(arguments[1]) * 0.5);
        var cosz = Math.cos(toRadians(arguments[2]) * 0.5);
        var sinx = Math.sin(toRadians(arguments[0]) * 0.5);
        var siny = Math.sin(toRadians(arguments[1]) * 0.5);
        var sinz = Math.sin(toRadians(arguments[2]) * 0.5);
        
        this.W = cosx * cosy * cosz - sinx * siny * sinz;
        this.X = sinx * cosy * cosz + cosx * siny * sinz;
        this.Y = siny * cosx * cosz - sinx * sinz * cosy;
        this.Z = cosx * cosy * sinz + sinx * siny * cosz;
        
        break;
      
      case 4:
        
        if(arguments[0] == 0 && arguments[1] == 0 && arguments[2] == 0 && arguments[3] == 0)
        {
          this.W = 1;
          this.X = 0;
          this.Y = 0;
          this.Z = 0;
        }
        else
        {
          this.W = arguments[0];
          this.X = arguments[1];
          this.Y = arguments[2];
          this.Z = arguments[3];
          
          var magnitude = Math.sqrt(Math.pow(this.W, 2) + Math.pow(this.X, 2) + Math.pow(this.Y, 2) + Math.pow(this.Z, 2));
          
          this.W = this.W/magnitude;
          this.X = this.X/magnitude;
          this.Y = this.Y/magnitude;
          this.Z = this.Z/magnitude;
        }
        
        break;
    }
  }
}

function multiplyQuat(quat1, quat2)
{
  var multipliedQuat = new quaternion();
  multipliedQuat.W = -quat1.X * quat2.X - quat1.Y * quat2.Y - quat1.Z * quat2.Z + quat1.W * quat2.W;
  multipliedQuat.X = quat1.X * quat2.W + quat1.Y * quat2.Z - quat1.Z * quat2.Y + quat1.W * quat2.X;
  multipliedQuat.Y = -quat1.X * quat2.Z + quat1.Y * quat2.W + quat1.Z * quat2.X + quat1.W * quat2.Y;
  multipliedQuat.Z = quat1.X * quat2.Y - quat1.Y * quat2.X + quat1.Z * quat2.W + quat1.W * quat2.Z;
  
  return multipliedQuat;
}

File: MapEditingTools/Quaternion.cs
using System;

namespace MapEditingTools
{
    public class Quaternion
    {
        public double W { get; set; }
        public double X { get; set; }
        public double Y { get; set; }
        public double Z { get; set; }

        public Quaternion()
        {
            W = 1;
            X = 0;
            Y = 0;
            Z = 0;
        }
        public Quaternion(double x, double y, double z)
        {
            double cosx = Math.Cos(ToRadians(x));
            double cosy = Math.Cos(ToRadians(y));
            double cosz = Math.Cos(ToRadians(z));
            double sinx = Math.Sin(ToRadians(x));
            double siny = Math.Sin(ToRadians(y));
            double sinz = Math.Sin(ToRadians(z));

            W = cosx * cosy * cosz - sinx * siny * sinz;
            X = sinx * cosy * cosz + cosx * siny * sinz;
            Y = siny * cosx * cosz - sinx * sinz * cosy;
            Z = cosx * cosy * sinz + sinx * siny * cosz;
        }
        public Quaternion(double w, double x, double y, double z)
        {
            if (w == 0 && x == 0 && y == 0 && z == 0)
            {
                W = 1;
                X = 0;
                Y = 0;
                Z = 0;
            }
            else
            {
                double magnitude = Math.Sqrt(Math.Pow(w, 2) + Math.Pow(x, 2) + Math.Pow(y, 2) + Math.Pow(z, 2));

                W /= magnitude;
                X /= magnitude;
                Y /= magnitude;
                Z /= magnitude;
            }
        }
        public static Quaternion MultiplyQuat(Quaternion quat1, Quaternion quat2)
        {
            Quaternion multipliedQuat = new Quaternion
            {
                W = -quat1.X * quat2.X - quat1.Y * quat2.Y - quat1.Z * quat2.Z + quat1.W * quat2.W,
                X = quat1.X * quat2.W + quat1.Y * quat2.Z - quat1.Z * quat2.Y + quat1.W * quat2.X,
                Y = -quat1.X * quat2.Z + quat1.Y * quat2.W + quat1.Z * quat2.X + quat1.W * quat2.Y,
                Z = quat1.X * quat2.Y - quat1.Y * quat2.X + quat1.Z * quat2.W + quat1.W * quat2.Z
            };

            return multipliedQuat;
        }
        public double ToRadians(double angle)
        {
            return (Math.PI / 180) *
                   (angle * 0.5); //Some Additional :D
        }
    }
}
